docs/decisions/0001-null-safety-strategy-for-operaton.md

```diff
@@ -0,0 +1,105 @@
+- Do not rely on preview/incubating language features.
+- Align with major ecosystem dependencies (e.g., Spring 7) to reduce friction for users.
+
+## Considered Options
+
+1. Status Quo — Keep `EnsureUtil` and `Objects.requireNonNull`, no nullability annotations.
+2. JSpecify for New Public APIs Only — Annotate only newly introduced public APIs going forward.
+3. JSpecify for All Public APIs (Selected) — Annotate all public-facing APIs; internals optional/gradual.
+4. JSpecify Project-Wide — Annotate public and internal codebase; highest migration cost.
+5. Defer Adoption — Revisit when JSpecify/Valhalla matures further.
+6. Use Another Annotation Framework (e.g., JetBrains) — Diverges from ecosystem direction; similar migration cost.
+
+## Decision Outcome
+
+Chosen option: "JSpecify for All Public APIs (Selected)", because it formalizes contracts at API boundaries, aligns with ecosystem direction and tooling, enables effective IDE/CI assistance, and allows gradual internal adoption with reasonable cost and risk.
+
+### Consequences
+
+- Good, because public API contracts become explicit (fewer “can this be null?” questions) and static analyzers/IDEs can detect contract violations earlier; alignment with ecosystem (e.g., Spring 7) reduces friction.
+- Bad, because contributors need light training; build/CI configuration must be adjusted; care is required to avoid behavior changes during refactors; partial coverage may exist during transition.
+
+### Confirmation
+
+Compliance will be confirmed via code review and CI checks:
+
+- 100% of public APIs annotated with JSpecify.
+- CI flags nullability violations on public APIs.
+- No externally visible behavior changes (e.g., exception types) in the engine module due to refactors.
+
+## Pros and Cons of the Options
+
+### Status Quo — Keep EnsureUtil/Objects.requireNonNull, no annotations
+
+- Good, because it has no immediate migration cost and preserves current behavior and CI setup.
+- Neutral, because runtime checks can still catch nulls at boundaries but provide no design-time guidance.
+- Bad, because ambiguity about nullability persists; limited tool-assisted safety; diverges from ecosystem direction.
+- Bad, because it adds runtime overhead.
+
+### JSpecify for New Public APIs Only
+
+- Good, because it minimizes churn and adds clarity for newly added surfaces.
+- Good, because runtime null checks can be reduced on hot paths to cut overhead.
+- Good, because null safety is part of the contract/API.
+- Neutral, because mixed annotation coverage can be acceptable temporarily.
+- Bad, because it leaves existing APIs ambiguous for a long time; inconsistent experience for users.
+- Bad, because of inconsistency in null safety across public APIs.
+
+### JSpecify for All Public APIs (Selected)
+
+- Good, because it sets clear, consistent contracts at API boundaries where they matter most; strong tool support; ecosystem alignment.
```

Comment:
Strong tool support is not a pro of this option, but for JSpecify itself. Therefore I would consider general JSpecify Pro/Cons and then discuss the scope.

Maybe the scope is even a separate ADR?